tiledb/sm/cpp_api/subarray.h

```diff
@@ -0,0 +1,695 @@
+   * @param dim_name The name of the dimension to add the range to.
+   * @param start The range start to add.
+   * @param end The range end to add.
+   * @param stride The range stride to add.
+   * @return Reference to this Query
+   */
+  template <class T>
+  Subarray& add_range(
+      const std::string& dim_name, T start, T end, T stride = 0) {
+    impl::type_check<T>(schema_.domain().dimension(dim_name).type());
+    auto& ctx = ctx_.get();
+    ctx.handle_error(tiledb_subarray_add_range_by_name(
+        ctx.ptr().get(),
+        subarray_.get(),
+        dim_name.c_str(),
+        &start,
+        &end,
+        (stride == 0) ? nullptr : &stride));
+    return *this;
+  }
+
+  /**
+   * Adds a 1D string range along a subarray dimension index, in the form
+   * (start, end). Applicable only to variable-sized dimensions
+   *
+   * **Example:**
+   *
+   * @code{.cpp}
+   * // Set a 1D range on dimension 0, assuming the domain type is int64.
+   * int64_t start = 10;
+   * int64_t end = 20;
+   * // Stride is optional
+   * subarray.add_range(0, start, end);
+   * @endcode
+   *
+   * @tparam T The dimension datatype
+   * @param dim_idx The index of the dimension to add the range to.
+   * @param start The range start to add.
+   * @param end The range end to add.
+   * @return Reference to this Query
+   */
+  Subarray& add_range(
+      uint32_t dim_idx, const std::string& start, const std::string& end) {
+    impl::type_check<char>(schema_.domain().dimension(dim_idx).type());
+    auto& ctx = ctx_.get();
+    ctx.handle_error(tiledb_subarray_add_range_var(
+        ctx.ptr().get(),
+        subarray_.get(),
+        dim_idx,
+        start.c_str(),
+        start.size(),
+        end.c_str(),
+        end.size()));
+    return *this;
+  }
+
+  /**
+   * Adds a 1D string range along a subarray dimension name, in the form (start,
+   * end). Applicable only to variable-sized dimensions
+   *
+   * **Example:**
+   *
+   * @code{.cpp}
+   * // Set a 1D range on dimension "rows", assuming the domain type is int64.
+   * int64_t start = 10;
+   * int64_t end = 20;
+   * const std::string dim_name = "rows";
+   * // Stride is optional
+   * subarray.add_range(dim_name, start, end);
+   * @endcode
+   *
+   * @tparam T The dimension datatype
+   * @param dim_name The name of the dimension to add the range to.
+   * @param start The range start to add.
+   * @param end The range end to add.
+   * @return Reference to this Query
+   */
+  TILEDB_DEPRECATED
+  Subarray& add_range(
+      const std::string& dim_name,
+      const std::string& start,
+      const std::string& end) {
+    impl::type_check<char>(schema_.domain().dimension(dim_name).type());
+    auto& ctx = ctx_.get();
+    ctx.handle_error(tiledb_subarray_add_range_var_by_name(
+        ctx.ptr().get(),
+        subarray_.get(),
+        dim_name.c_str(),
+        start.c_str(),
+        start.size(),
+        end.c_str(),
+        end.size()));
+    return *this;
+  }
+
+  /**
```

Comment:
we shouldn't need this routine, the caller can use `ptr()`

Reply:
done